### PR TITLE
JOINDIN-648 allow different talk types

### PR DESCRIPTION
**Pull request by exussum12**
Added new backend access for different talk type medias

This adds the new data to the verbose getTalk call and also added a new endpoint for getting all of the media types.

/v2.1/talks/3/links

and 
/v2.1/talks/3?verbose=yes

Have changed.

There is a backwards compatible change for for slide links (when they get moved over with another ticket)

### CODE REVIEW

File: db/patch68.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
+CREATE TABLE `talk_types_link` (
```
Comment (exussum12) [inline db/patch68.sql]: Not sure if this is the correct way to do DB changes ? Its basically these two tables

File: src/controllers/TalksController.php [hunk 2]
```diff
@@ -51,13 +51,11 @@ public function getAction($request, $db)
 
     public function postAction($request, $db)
     {
-        if (! isset($request->user_id)) {
-            throw new Exception("You must be logged in to create data", 401);
-        }
+        $this->checkLoggedIn($request);
```
Comment (exussum12) [inline src/controllers/TalksController.php]: Refacored all of these, They are everywhere.

File: src/controllers/TalksController.php [hunk 3]
```diff
@@ -51,13 +51,11 @@ public function getAction($request, $db)
...
         $talk_id = $this->getItemId($request);
 
         // Retrieve the talk. It if doesn't exist, then 404 with talk not found
-        $talk= $this->getTalkById($db, $request, $talk_id);
+        $talk = $this->getTalkById($request, $db, $talk_id);
```
Comment (exussum12) [inline src/controllers/TalksController.php]: Argument order switched for consistency

File: src/controllers/TalksController.php [hunk 4]
```diff
@@ -215,16 +211,11 @@ public function deleteAction($request, $db)
      */
     public function addTrackToTalk(Request $request, PDO $db)
     {
-        if (!isset($request->user_id)) {
-            throw new Exception("You must be logged in to create data", 400);
```
Comment (exussum12) [inline src/controllers/TalksController.php]: Oddly a few of the existing requests had 400 response for this ? I had changed all of these to 401. Possible breaking change

File: src/controllers/TalksController.php [hunk 5]
```diff
@@ -822,16 +788,46 @@ public function removeSpeakerForTalk(Request $request, PDO $db)
             throw new Exception("There was a problem assigning the talk", 500);
         }
 
-        //If we are unit testing, then we can't exit or send headers!
-        if (defined('UNIT_TEST')) {
```
Comment (exussum12) [inline src/controllers/TalksController.php]: adjusted these to match @heiglandreas  change. His was merged while this pull request was open.

File: src/models/OAuthModel.php [hunk 6]
```diff
@@ -167,11 +167,9 @@ public function newAccessToken($consumer_key, $user_id)
     {
         $hash              = $this->generateToken();
         $accessToken       = substr($hash, 0, 16);
-        $accessTokenSecret = substr($hash, 16, 16);
 
         $sql = "INSERT INTO oauth_access_tokens set
                 access_token = :access_token,
-                access_token_secret = :access_token_secret,
```
Comment (exussum12) [inline src/models/OAuthModel.php]: This was never used, Removed

File: tests/controllers/TalksControllerTest.php [hunk 7]
```diff
@@ -18,10 +23,16 @@ class TalksControllerTest extends \PHPUnit_Framework_TestCase
      */
     public function testClaimTalkWithNoUserIdThrowsException()
     {
-        $request = new \Request([], ['REQUEST_URI' => "http://api.dev.joind.in/v2.1/talks/326/speakers", 'REQUEST_METHOD' => 'POST']);
+        $request = new Request(
```
Comment (exussum12) [inline tests/controllers/TalksControllerTest.php]: These were all > 120. Reduced to < 120 for PSR-2

File: tests/controllers/TalksControllerTest.php [hunk 8]
```diff
@@ -33,17 +44,23 @@ public function testClaimTalkWithNoUserIdThrowsException()
      *
      * @return void
      *
-     * @test
```
Comment (exussum12) [inline tests/controllers/TalksControllerTest.php]: Removed a few of these too - all of the functions are prefixed with test so seems pointless to have the test annotation too

File: db/patch68.sql (hunk 1 above)
Comment (heiglandreas) [inline db/patch68.sql]: Are we talking about talk_types or slides?

File: db/patch68.sql [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+CREATE TABLE `talk_types` (
+  `ID` int(11) NOT NULL AUTO_INCREMENT,
+  `type` varchar(255) NOT NULL,
+  `display_name` varchar(255) DEFAULT NULL,
+  PRIMARY KEY (`ID`)
+) ENGINE=InnoDB;
+
+
+INSERT INTO `talk_types`(`ID`,`type`,`display_name`) VALUES
+(1,'Slides','slides_link'),
```
Comment (heiglandreas) [inline db/patch68.sql]: Shouldn't that be (1, 'slides_link', 'Slides')?

File: db/patch68.sql [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+(2,'Video','video_link'),
+(3,'Audio','audio_link');
+
```
Comment (heiglandreas) [inline db/patch68.sql]: At the end you should have a line ```INSERT INTO patch_history SET patch_number = 68;```

File: src/controllers/TalksController.php (hunk 2 above)
Comment (heiglandreas) [inline src/controllers/TalksController.php]: I'd prefer to have those refactorings in a separate PR. They aren't necessary for the feature. Having them in two separate PRs makes each one easier to digest :)

File: db/patch68.sql (hunk 1 above)
Comment (exussum12) [inline db/patch68.sql]: Slides is a form of talk type. Though agreed the name could be better. Any suggestions ?

File: db/patch68.sql (hunk 9 above)
Comment (exussum12) [inline db/patch68.sql]: display_name here is the display name when looking at the endpoint. Its currently slides_link and should not really change (as the front end would need some additional code to handle it)

File: db/patch68.sql (hunk 10 above)
Comment (exussum12) [inline db/patch68.sql]: Will add that thanks :)

File: db/patch68.sql (hunk 1 above)
Comment (heiglandreas) [inline db/patch68.sql]: Aren't they talk_links?

File: src/controllers/TalksController.php (hunk 4 above)
Comment (heiglandreas) [inline src/controllers/TalksController.php]: We had a discussion about that a while back and decided to leave them at the moment as they are. If changed that should also be a separate PR that then changes *all* of those occurences throughout the API

File: src/controllers/TalksController.php (hunk 5 above)
Comment (heiglandreas) [inline src/controllers/TalksController.php]: This also (despite being a good one) should also be a separate PR.

File: src/models/OAuthModel.php (hunk 6 above)
Comment (heiglandreas) [inline src/models/OAuthModel.php]: Is that necessary for the talk links? Otherwise please add that as a further PR. 

And even though it's not used ATM it might become necessary in the future.

File: tests/controllers/TalksControllerTest.php (hunk 7 above)
Comment (heiglandreas) [inline tests/controllers/TalksControllerTest.php]: Also please add this as a separate PR.

File: tests/controllers/TalksControllerTest.php (hunk 8 above)
Comment (heiglandreas) [inline tests/controllers/TalksControllerTest.php]: Though you are absolutely right, that's a refactoring that should also go into it's own PR

File: src/controllers/TalksController.php [hunk 11]
```diff
@@ -215,16 +211,11 @@ public function deleteAction($request, $db)
...
-        if (!isset($request->user_id)) {
-            throw new Exception("You must be logged in to create data", 400);
-        }
+        $this->checkLoggedIn($request);
 
-        $talk_id = $this->getItemId($request);
         $talk_mapper = new TalkMapper($db, $request);
-        $talk = $talk_mapper->getTalkById($talk_id);
-        if (!$talk) {
-            throw new Exception("Talk not found", 404);
```
Comment (heiglandreas) [inline src/controllers/TalksController.php]: Why do you need to remove this line? Is that relevant to the talk-links?

Removing it has two downsides: 

1. line 218 might fail when the user asks for a talk not available
2. The user will never know that the talk doesn't exist.

File: src/controllers/TalksController.php [hunk 12]
```diff
@@ -268,18 +259,13 @@ public function addTrackToTalk(Request $request, PDO $db)
...
-        }
+        $this->checkLoggedIn($request);
 
-        $talk_id = $this->getItemId($request);
         $track_id = $request->url_elements[5];
 
         $talk_mapper = new TalkMapper($db, $request);
-        $talk = $talk_mapper->getTalkById($talk_id);
-        if (!$talk) {
-            throw new Exception("Talk not found", 404);
```
Comment (heiglandreas) [inline src/controllers/TalksController.php]: see above…

File: src/models/OAuthModel.php (hunk 6 above)
Comment (exussum12) [inline src/models/OAuthModel.php]: Then add it in the future when it *is* needed. Keeping dead code in the codebase is silly

Comment (heiglandreas) [inline src/models/OAuthModel.php]: It's already there. And removing it is not the purpose of this PR.

File: src/controllers/TalksController.php [hunk 13]
```diff
@@ -320,10 +306,18 @@ public function removeTrackFromTalk(Request $request, PDO $db)
      *
      * @return TalkModelCollection
      */
-    protected function getTalkById($db, $request, $talk_id)
-    {
-        $mapper = new TalkMapper($db, $request);
-        $talk   = $mapper->getTalkById($talk_id);
+    protected function getTalkById(
```
Comment (heiglandreas) [inline src/controllers/TalksController.php]: Rewriting this method is not necessary for the scope of the PR. so please add it as a separate PR.

File: db/patch68.sql (hunk 9 above)
Comment (heiglandreas) [inline db/patch68.sql]: Then what is the type needed for? A display name in my eyes is something that should be displayed to an end user. Have a look at https://github.com/joindin/joindin-api/blob/master/src/controllers/LanguagesController.php for an example

File: src/models/OAuthModel.php [hunk 14]
```diff
@@ -181,7 +179,6 @@ public function newAccessToken($consumer_key, $user_id)
         $result = $stmt->execute(
             array(
                 'access_token'        => $accessToken,
-                'access_token_secret' => $accessTokenSecret,
```
Comment (heiglandreas) [inline src/models/OAuthModel.php]: See above.

File: tests/controllers/TalksControllerTest.php [hunk 15]
```diff
@@ -775,6 +861,50 @@ public function testRejectClaimAsHostSucceeds()
...
+            ]
+        );
+        $request->user_id = 1;
+        $request->parameters = [
+            'username'      => 'psherman',
+            'display_name'  => 'P Sherman',
+            'verbose'       => 'yes',
+        ];
+        $db = $this->getMockBuilder(mockPDO::class)->getMock();
+        $stmt = $this->getMockBuilder(PDOStatement::class)
```
Comment (heiglandreas) [inline tests/controllers/TalksControllerTest.php]: It would be better to inject a mocked TalkMediaMapper into the controller. That way you can test the controller independent from the mapper. That would then also require tests for the TalkMediaMapper.

File: db/patch68.sql (hunk 9 above)
Comment (exussum12) [inline db/patch68.sql]: Should an API have any say at all on the front end appearance ?

File: src/controllers/TalksController.php (hunk 13 above)
Comment (exussum12) [inline src/controllers/TalksController.php]: It is needed for the verbose element to be added

File: tests/controllers/TalksControllerTest.php (hunk 15 above)
Comment (exussum12) [inline tests/controllers/TalksControllerTest.php]: Good shout. Was testing too much. I have moved that now

File: db/patch68.sql (hunk 10 above)
Comment (exussum12) [inline db/patch68.sql]: added thanks

File: src/controllers/TalksController.php (hunk 4 above)
Comment (exussum12) [inline src/controllers/TalksController.php]: Changed all of them back to throwing 400's (which did previously)

File: db/patch68.sql (hunk 1 above)
Comment (exussum12) [inline db/patch68.sql]: Changed the names.

**Comment by exussum12** [comment]
Not sure if we want to just move this to the other PR ? Conflict resolved and as it built on this contains both

**Comment by exussum12** [comment]
Closing due to all of this being contained in the other PR
